### DISCUSSION TITLE
[Maui] How init essentials?

### DISCUSSION THREAD

**Discussion by mcocaroUX** [unoplatform/uno#15651] [Q&A] [answered]
We're currently developing a payments feature in our UNO Platform app and have chosen to use the [InAppBillingPlugin](https://github.com/jamesmontemagno/InAppBillingPlugin) by James Montemagno. This plugin needs the use of Maui.Essentials for its operation. Although the installation of the plugin and Maui.Essentials proceeded without issues, and the app runs, we're encountering difficulties with correctly initializing Maui.Essentials. This initialization issue is subsequently causing failures in the Billing plugin functionality.

We followed the standard initialization procedure for Maui.Essentials as documented, but it seems like the plugin is not recognizing this initialization, or perhaps we're missing a step specific to the UNO Platform context. Given the UNO Platform's unique architecture, integrating Maui.Essentials might have nuances that we're not fully aware of.

Any guidance or suggestions on how to troubleshoot this problem would be greatly appreciated.

Thank you in advance for your help

**Comment by pictos** [comment]
@mcocaroUX do you mind sharing the issues that you're facing? Like error messages, or the current value and expected value (Android activity should be XXX but it's null), etc

**Reply by mcocaroUX** [reply]
This is the error message I'm getting: `Current Activity is null, ensure that the MainActivity.cs file is configuring Xamarin.Essentials/.NET MAUI in your source code so the In App Billing can use it.`

**Reply by pictos** [reply]
Yeah, I saw that before. I've a workaround let me see if it still works.

Just to put some light on the problem, the way Maui and Uno initialize the app are different, so the Maui.Embedding doesn't have info about the Activity, so the Essentials (that are initialized on Maui.Builder) doesn't know anything about the Activity created by Uno

**Reply by pictos** [reply]
@mcocaroUX what version of Uno packages are you using? I'm trying with the latest and I was able to get the `Platform.CurrentActivity` from Essentials

**Reply by mcocaroUX** [reply]
UnoVersion = 5.1.75

```
  <ItemGroup>
    <PackageVersion Include="Microsoft.Extensions.Logging.Console" Version="8.0.0" />
    <PackageVersion Include="Microsoft.Windows.SDK.BuildTools" Version="10.0.22621.2428" />
    <PackageVersion Include="Microsoft.WindowsAppSDK" Version="1.4.231219000" />
    <PackageVersion Include="Uno.Core.Extensions.Logging.Singleton" Version="4.0.1" />
    <PackageVersion Include="Uno.Extensions.Authentication" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Authentication.WinUI" Version="4.0.0" />

    <PackageVersion Include="Uno.Extensions.Configuration" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Hosting" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Hosting.WinUI" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Http" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Http.WinUI" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Http.Refit" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Localization" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Localization.WinUI" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Logging.OSLog" Version="1.7.0" />
    <PackageVersion Include="Uno.Extensions.Logging.WinUI" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Navigation" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Navigation.WinUI" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Navigation.Toolkit.WinUI" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Reactive" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Reactive.WinUI" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Serialization" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Serialization.Http" Version="4.0.0" />
    <PackageVersion Include="Uno.Extensions.Serialization.Refit" Version="4.0.0" />
    <PackageVersion Include="Uno.Material.WinUI" Version="4.1.0" />
    <PackageVersion Include="Uno.Dsp.Tasks" Version="1.2.8" />
    <PackageVersion Include="Uno.Toolkit.WinUI" Version="5.1.5" />
    <PackageVersion Include="Uno.Toolkit.WinUI.Material" Version="5.1.5" />
    <PackageVersion Include="Uno.Resizetizer" Version="1.2.1" />

    <PackageVersion Include="Uno.UI.Adapter.Microsoft.Extensions.Logging" Version="$(UnoVersion)" />
    <PackageVersion Include="Uno.UniversalImageLoader" Version="1.9.36" />
    <PackageVersion Include="Uno.WinUI" Version="$(UnoVersion)" />
    <PackageVersion Include="Uno.WinUI.Lottie" Version="$(UnoVersion)"/>
    <PackageVersion Include="Uno.WinUI.RemoteControl" Version="$(UnoVersion)" />
    <PackageVersion Include="Xamarin.Google.Android.Material" Version="1.9.0.2" />
    <PackageVersion Include="Mapsui.Uno.WinUI" Version="4.1.2" />
    <PackageVersion Include="Uno.Extensions.Reactive.Messaging" Version="2.5.11" />
    <PackageVersion Include="CommunityToolkit.Mvvm" Version="8.2.0" />
    <PackageVersion Include="Uno.WinUI.Svg" Version="$(UnoVersion)" />
    <PackageVersion Include="SkiaSharp.Views.Uno" Version="2.88.6" />
    <PackageVersion Include="Xamarin.Android.Google.BillingClient" Version="6.0.1.4" />
    <PackageVersion Include="Xamarin.Android.Google.BillingClient.Ktx" Version="6.0.1.4" />
    <PackageVersion Include="Xamarin.AndroidX.Legacy.Support.V4" Version="1.0.0.18" />

    <PackageVersion Include="Uno.WinUI.DevServer" Version="$(UnoVersion)" />
    <PackageVersion Include="Plugin.InAppBilling" Version="7.1.0" />
    <PackageVersion Include="Microsoft.Maui.Essentials" Version="8.0.3" />
  </ItemGroup>
```

**Reply by mcocaroUX** [reply]
@pictos I updated all the packaged and couldn't get it to work, what is the workaround you mentioned?

**Reply by dansiegel** [reply, marked as answer]
In your mobile application under Android... open the Main.Android... this gives you the native Android Application. In the constructor add this line:

```cs
Microsoft.Maui.ApplicationModel.Platform.Init(this);
```

This will initialize Maui Essentials for Android and hook up the ActivityStateManager that returns the current Activity.
